libpkg/elfhints.c

```diff
@@ -66,8 +66,11 @@ static void	write_elf_hints(const char *);
 
 static const char	*dirs[MAXDIRS];
 static int		 ndirs;
+static int		 is_be;
 int			 insecure;
 
+#define COND_SWAP(n) (is_be ? be32toh(n) : le32toh(n))
```

Comment:
I'm not expert in this code... but shouldn't byte-order be determined at compile time, for example, (`#if __BYTE_ORDER__ == __BIG_ENDIAN__`)? As on LE, `le32toh()` is just a cast to `(uint32_t)` so the work is only needed for big endian architectures...

Reply:
I have no read the code yet, but this should be dynamically determined in pkg as pkg can cross install packages:
pkg -r /rootdir install something

this something can be of a different endianness

Reply:
I see that make sense, thanks!

Reply:
The goal is to make the hints file little-endian an FreeBSD on all architectures. If you look at line 2111 of rtld.c in 
https://reviews.freebsd.org/D44053 , you'll see that there is a commented out test for being built on a little-endian host that implements the suggested optimization:

`	is_le = /*le32toh(1) == 1 || */ hdr.magic == ELFHINTS_MAGIC;`

It is possible to restrict ldconfig (and pkg) to little-endian hints files at compile time on little-endian hosts, but the cost of the extra test and the overhead is extremely low compared to the cycles spent by the pkg command for its general operation.

Therefore, I'd prefer the flexibility offered by dynamically adapting to the byte-order of an existing hints file, and thus allow updating a big-endian hints file when executed on a little-endian host.
This would for example allow to install packages for a file system prepared for a powerpc64 emulation using the amd64 pkg binary on an amd64 host.

If such a flexibility is deemed unnecessary, I'll update the patch to be optimized for little-endian machines (and remove their big-endian hints file support). I prefer the expression quoted above to a test of __BYTE_ORDER__, though, modified for is_be as used in this pull request:

`	is_be = be32toh(1) == 1 || be32toh(hdr->magic) == ELFHINTS_MAGIC;`

(In fact, I'm currently working on an amd64 system with a big-endian hints file to test these changes in all affected commands (rtld, ldconfig, and pkg) ... This is equivalent to testing on e.g. a powerpc64 host with little-endian hints file due the symmetry of the respective byte-swap operations.)